extensions/storages/source/providers/S3.ts

```diff
@@ -0,0 +1,107 @@
+import { type Readable } from 'node:stream'
+import { join } from 'node:path/posix'
+import { Upload } from '@aws-sdk/lib-storage'
+import dotenv from 'dotenv'
+import fse from 'fs-extra'
+import {
+  S3Client,
+  GetObjectCommand,
+  CopyObjectCommand,
+  ListObjectsV2Command,
+  DeleteObjectsCommand,
+  type S3ClientConfigType
+} from '@aws-sdk/client-s3'
+import { type Provider } from '../Provider'
+
+if (fse.existsSync(join(__dirname, '.env')))
+  dotenv.config({ path: join(__dirname, '.env') })
+
+const {
+  S3_ACCESS_KEY = '',
+  S3_SECRET_ACCESS_KEY = ''
+} = process.env
+
+export class S3 implements Provider {
+  protected readonly bucket: string
+  protected readonly client: S3Client
+
+  public constructor (url: URL) {
+    const pathSegments = url.pathname.split('/').filter((segment) => segment !== '')
+
+    if (pathSegments.length === 0) throw new Error('S3 URL must contain at least two segments')
+
+    const s3Config: S3ClientConfigType = {
+      credentials: {
+        accessKeyId: S3_ACCESS_KEY,
+        secretAccessKey: S3_SECRET_ACCESS_KEY
+      }
+    }
+
+    if (pathSegments.length === 1) {
+      s3Config.region = url.host
+      this.bucket = pathSegments[0]
+    } else {
+      s3Config.region = pathSegments[0]
+      s3Config.endpoint = `http://${url.host}` // ???
```

Comment:
@temich, maybe we should use environment variables for the `endpoint`? The SDK doesn't accept an `endpoint` without a protocol.